Add documentation page for checks and assertions (#33241)
hashicorp/terraform: website/docs/language/expressions/custom-conditions.mdx

> **Hands On:** Try the [Validate Modules with Custom Conditions](/terraform/tutorials/configuration-language/custom-conditions) tutorial.

This page explains the following:
  - Creating checks with [assertions](#check-assertions) to verify your infrastructure as a whole (Terraform v1.5.0 and later)
  - Creating [validation conditions](#input-variable-validation) for input variables (Terraform v0.13.0 and later)
  - Creating [preconditions and postconditions](#preconditions-and-postconditions) for resources, data sources, and outputs (Terraform v1.2.0 and later)
  - Writing effective [condition expressions](#condition-expressions) and [error messages](#error-messages)
  - When Terraform [evaluates custom conditions](#conditions-checked-only-during-apply) during the plan and apply cycle

## Selecting a Custom Condition for Your Use Case

Terraform's different custom conditions are best suited to various situations. Use the following broad guidelines to select the best custom condition for your use case:

> 1. [Check blocks with assertions](#check-assertions) validate your infrastructure as a whole. In addition, check blocks are non-invasive and so do not prevent or block the overall execution of Terraform operations.

Reviewer: ```suggestion
1. [Check blocks with assertions](#check-assertions) validate your infrastructure as a whole. Additionally, check blocks do not prevent or block the overall execution of Terraform operations.
```

This page explains the following:
  - Creating [validation conditions](#input-variable-validation) for input variables (Terraform v0.13.0 and later)
  - Creating [preconditions and postconditions](#preconditions-and-postconditions) for resources, data sources, and outputs (Terraform v1.2.0 and later)
  - Creating [assertions](#check-assertions) within check blocks (Terraform v1.5.0 and later)
  - Writing effective [condition expressions](#condition-expressions) and [error messages](#error-messages)
  - When Terraform [evaluates custom conditions](#conditions-checked-only-during-apply) during the plan and apply cycle

> 

Reviewer: Below here, I think it'd be nice to have a paragraph clarifying when users should use which kind of custom condition. Especially since folks can do things in difference ways, we want **strong** guidance at the beginning so folks don't have to read this whole page to figure out which kind they should use (checks vs. pre/post conditions .v.s. validation blocks). 

This paragraph doesn't need to describe how the feature works, we can link to sections below.

This page explains the following:
  - Creating [validation conditions](#input-variable-validation) for input variables (Terraform v0.13.0 and later)
  - Creating [preconditions and postconditions](#preconditions-and-postconditions) for resources, data sources, and outputs (Terraform v1.2.0 and later)
  - Creating [assertions](#check-assertions) within check blocks (Terraform v1.5.0 and later)
  - Writing effective [condition expressions](#condition-expressions) and [error messages](#error-messages)
  - When Terraform [evaluates custom conditions](#conditions-checked-only-during-apply) during the plan and apply cycle

> The various custom conditions are suited to different situations. Input variable validation and output postconditions should be used to validate the inputs and outputs of your configuration meet any requirements or conditions. Resource preconditions and postconditions should be used to validate your configuration is produced expected and predictable results. Check blocks and assertions should be used to evaluate your infrastructure as a whole.

Reviewer: Now that I look at it, it feels like this  paragraph should be it's own section? How do you feel about something like the following? I also reformatted this paragraph as a list because it felt easier to parse, and rearranged the order of options to feature checks first because it one of the broadest use cases:

```suggestion
## Selecting a Custom Condition for Your Use Case

Terraform's different custom conditions are best suited to various situations. Use the following broad guidelines to select the best custom condition for your use case:
1. [Check blocks with assertions](#check-assertions) validate your infrastructure as a whole.
1. [Validation conditions](#input-variable-validation) or [output postconditions](#preconditions-and-postconditions) can ensure your configuration's inputs and outputs meet specific requirements.
1. Resource [preconditions and postconditions](#preconditions-and-postconditions) can validate that Terraform produces your configuration with predictable results.
```

> For more discussion on when to use different custom conditions, please read [Choosing Between Preconditions and Postconditions](#choosing-between-preconditions-and-postconditions) and [Choosing Checks or other Custom Conditions](/terraform/language/checks#choosing-checks-or-other-custom-conditions).

Reviewer: Tiny tweak for style:
```suggestion
For more information on when to use certain custom conditions, see [Choosing Between Preconditions and Postconditions](#choosing-between-preconditions-and-postconditions) and [Choosing Checks or Other Custom Conditions](/terraform/language/checks#choosing-checks-or-other-custom-conditions).
```

- Which approach is more convenient? If a particular resource has many dependencies that all make an assumption about that resource, it can be pragmatic to declare that once as a post-condition of the resource, rather than declaring it many times as preconditions on each of the dependencies.
- Is it helpful to declare the same or similar conditions as both preconditions and postconditions? This can be useful if the postcondition is in a different module than the precondition because it lets the modules verify one another as they evolve independently.

> ## Check Assertions

Author: So in this case what I want to refer to is the `assert` statements within the check blocks. This is because the assert block within a check block matches the syntax for the other custom conditions (pre/postconditions, input validations) discussed in this page.

Pending the other rename (Checks to Check blocks), we could rename this to Check block assertions instead? I would like to highlight that we are specifically talking about the assert sub block somehow.

-> **Note:** Check blocks and their assertions are only available in Terraform v1.5.0 and later.

> 

Reviewer: Can we add more flavor text to the beginning of this? A broader, "this is when and why you use checks" paragraph?

Reviewer: Also please mention when checks are actually run!

Reviewer: I think this is the perfect amount of context, thank you so much for adding it!

My general rule of thumb is usually, "Repeat information **if** someone needs more context to understand what we are talking about on _this_ page." I like to keep things as DRY as possible, but also ensure folks have enough context _without_ having to open to a new page. It's _definitely_ a thin line to walk though 😂

> Add one or more `assert` blocks within a `check` block to specify custom conditions. Each assertion requires a [`condition` argument](#condition-expressions), an expression that must use the value of the variable to return `true` if the value is valid, or `false` if it is invalid. The expression can refer to any resource, data source, or variable available to the surrounding check block.

Reviewer: (ripped this text straight from the condition section below. If you'd like to customize it further feel free! 
```suggestion
You can add one or more `assert` blocks within a `check` block to verify custom conditions. Each assertion requires a [`condition` argument](#condition-expressions), a boolean expression that should return `true` if the intended assumption or guarantee is fulfilled or `false` if it does not. Your `condition` expression can refer to any resource, data source, or variable available to the surrounding `check` block.
```

> 

Reviewer: Can you please add a code block so folks can see the terms you are referencing?

If the condition evaluates to `false`, Terraform produces an [error message](#error-messages) that includes the result of the `error_message` expression. If you declare multiple assertions, Terraform returns error messages for all failed conditions.

For more information on check blocks in general, refer to the Terraform language page for [Checks](/terraform/language/checks).

> ## Condition Expressions

Reviewer: We should update the sentence on the line _right_ underneath this one to include "checks":

"Checks, input variable validation, preconditions, and postconditions all require a condition argument."

Reviewer: Underneath this section is one named ["Conditions Checked Only During Apply",](https://developer.hashicorp.com/terraform/language/expressions/custom-conditions#conditions-checked-only-during-apply) does it need to be updated for Checks? It feels like yes, but wanted to see what you thought.
